extension/jaegerremotesampling/internal/http.go

```diff
@@ -0,0 +1,122 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/jaegerremotesampling/internal"
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/jaegertracing/jaeger/cmd/agent/app/configmanager"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/confighttp"
+)
+
+var (
+	errMissingClientConfigManager = errors.New("the client config manager has not been provided")
+)
+
+var _ component.Component = (*SamplingHTTPServer)(nil)
+
+type SamplingHTTPServer struct {
+	telemetry component.TelemetrySettings
+	settings  confighttp.HTTPServerSettings
+	cfgMgr    configmanager.ClientConfigManager
+
+	mux        *http.ServeMux
+	srv        *http.Server
+	shutdownWG *sync.WaitGroup
+}
+
+func NewHTTP(telemetry component.TelemetrySettings, settings confighttp.HTTPServerSettings, cfgMgr configmanager.ClientConfigManager) (*SamplingHTTPServer, error) {
+	if cfgMgr == nil {
+		return nil, errMissingClientConfigManager
+	}
+
+	srv := &SamplingHTTPServer{
+		telemetry: telemetry,
+		settings:  settings,
+		cfgMgr:    cfgMgr,
+
+		shutdownWG: &sync.WaitGroup{},
+	}
+
+	srv.mux = http.NewServeMux()
```

Comment:
I need it during the shutdown as well.

Reply:
True, but I had two things in mind:

1. I can use the mux in the tests, ensuring that the HTTP handler that I built has the right endpoints wired in. At the `Start`, I would have to provide a way to replace the HTTP server with the testing one.
2. It feels like returning a ready-to-use HTTP handler here is the right thing to do, instead of relying on the Start to complete the data structure *and* start it.

Reply:
I'm leaving this as is, but will change it if you don't agree with my points from the previous message.